src/ore/src/region.rs

```diff
@@ -148,3 +150,232 @@ impl<T> LgAllocRegion<T> {
+        }
+    }
+
+    /// Returns the capacity of the underlying allocation.
+    #[inline]
+    #[must_use]
+    pub fn capacity(&self) -> usize {
+        match self {
+            Region::Heap(vec) => vec.capacity(),
+            Region::MMap(inner) => inner.inner.capacity(),
+        }
+    }
+
+    /// Returns the number of elements in the allocation.
+    #[inline]
+    #[must_use]
+    pub fn len(&self) -> usize {
+        match self {
+            Region::Heap(vec) => vec.len(),
+            Region::MMap(inner) => inner.len(),
+        }
+    }
+
+    /// Returns true if the region does not contain any elements.
+    #[inline]
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        match self {
+            Region::Heap(vec) => vec.is_empty(),
+            Region::MMap(inner) => inner.is_empty(),
+        }
+    }
+
+    /// Dereference to the contained vector
+    #[inline]
+    #[must_use]
+    pub fn as_vec(&self) -> &Vec<T> {
+        match self {
+            Region::Heap(vec) => vec,
+            Region::MMap(inner) => &inner.inner,
+        }
+    }
+
+    /// Extend the underlying region from the iterator.
+    ///
+    /// Care must be taken to not re-allocate the inner vector representation.
+    #[inline]
+    pub fn extend<I: IntoIterator<Item = T> + ExactSizeIterator>(&mut self, iter: I) {
+        assert!(self.capacity() - self.len() >= iter.len());
+        // SAFETY: We just asserted that we have sufficient capacity.
+        unsafe { self.as_mut_vec().extend(iter) };
+    }
+
+    /// Obtain a mutable reference to the inner vector representation.
+    ///
+    /// Unsafe because the caller has to make sure that the vector will not reallocate.
+    #[inline]
+    pub unsafe fn as_mut_vec(&mut self) -> &mut Vec<T> {
+        match self {
+            Region::Heap(vec) => vec,
+            Region::MMap(inner) => &mut inner.inner,
+        }
+    }
+}
+
+impl<T: Clone> Region<T> {
+    /// Extend the region from a slice.
+    #[inline]
+    pub fn extend_from_slice(&mut self, slice: &[T]) {
+        assert!(self.capacity() - self.len() >= slice.len());
+        // SAFETY: We just asserted that we have enough capacity.
+        unsafe { self.as_mut_vec() }.extend_from_slice(slice);
+    }
+}
+
+impl<T> Deref for Region<T> {
+    type Target = [T];
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_vec()
+    }
+}
+
+impl<T> DerefMut for Region<T> {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        // SAFETY: We're dereferencing to `&mut [T]`, which does not allow reallocating the
+        // underlying allocation, which makes it safe.
+        unsafe { self.as_mut_vec().as_mut_slice() }
+    }
+}
+
+impl<T> Drop for Region<T> {
+    #[inline]
+    fn drop(&mut self) {
+        match self {
+            Region::Heap(vec) => {
+                // SAFETY: Don't drop the elements, drop the vec.
+                unsafe { vec.set_len(0) }
```

Comment:
Can you add a note here that we're purposefully choosing not to drop the elements for consistency?

Reply:
Added a comment.